crates/protocol/build.rs

```diff
@@ -28,3 +30,32 @@ fn main() -> Result<()> {
 
     Ok(())
 }
+
+// NB: protoc changed versioning schemes with v3.20 (removed major version), see https://protobuf.dev/support/version-support/
+// This means that after 3.20.x came 21.x - the major and minor version arguments here refer to the new scheme and would have previously been minor and patch versions respectively.
+fn verify_protoc_version(min_major: u32, min_minor: u32) {
+    let protoc_path = prost_build::protoc_from_env();
+    let protoc_version_output = Command::new(protoc_path)
+        .args(["--version"])
+        .output()
+        .unwrap();
+    let protoc_version = String::from_utf8(protoc_version_output.stdout).unwrap();
+
+    // based on semver.org's recommended regex, modified for lib name and optional patch capture group to accomodate libprotoc's versioning scheme
+    let re = Regex::new(r"^libprotoc (?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.?(?P<patch>0|[1-9]\d*)?(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$").unwrap();
```

Comment:
Any reason to hard-code a regex instead of using [`semver::Version::parse`](https://docs.rs/semver/1.0.17/semver/struct.Version.html#method.parse)?

Reply:
Yes, namely encountering the very first error case mentioned in the link you gave ;) Though I admit I actually assumed it simply didn't support fewer than three segments full stop - the docs seem to indicate that one could use VersionReq instead. Since semver is already in the deps anyway, I'll give that a try and see if it's a little more elegant.

Reply:
Interesting, it looks like according to [semver.org](https://semver.org/) the `semver` crate is correct.

```bnf
<version core> ::= <major> "." <minor> "." <patch>
```

Reply:
That's sort of a moot point though, we do need to parse these whether they count as semantic or not and if we need a regex that's fine.